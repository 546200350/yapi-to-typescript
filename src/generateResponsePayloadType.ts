import { JSONSchema4 } from 'json-schema'
import JSON5 from 'json5'
import { Api, ResponseBodyType } from './types'
import mockjsJsonToJsonSchema from './mockjsJsonToJsonSchema'
import jsonSchemaToTypes from './jsonSchemaToTypes'

/**
 * 生成返回数据的 typescript 类型定义。
 *
 * @param api Api
 * @param interfaceName 接口名称
 * @param [dataKey] 数据所在字段，不设置表示整体都是数据
 * @returns typescript 定义
 */
export default async function generateResponsePayloadType(api: Api, interfaceName: string, dataKey?: string): Promise<string> {
  let jsonSchema: JSONSchema4
  switch (api.res_body_type) {
    case ResponseBodyType.Json:
    case ResponseBodyType.JsonSchema:
      if (api.res_body_is_json_schema) {
        jsonSchema = api.res_body && JSON.parse(api.res_body)
      } else {
        jsonSchema = api.res_body && mockjsJsonToJsonSchema(JSON5.parse(api.res_body))
      }
      break
    default:
      break
  }
  if (dataKey && jsonSchema && jsonSchema.properties && jsonSchema.properties[dataKey]) {
    jsonSchema = jsonSchema.properties[dataKey]
  }
  return jsonSchemaToTypes(jsonSchema, interfaceName)
}
