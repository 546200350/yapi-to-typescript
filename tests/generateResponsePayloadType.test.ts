import generateResponsePayloadType from '../src/generateResponsePayloadType'

test('mockjs-json', async () => {
  const api: any = {
    _id: 1460,
    method: 'POST',
    catid: 91,
    title: 'post',
    path: '/post',
    project_id: 27,
    res_body_type: 'json',
    uid: 63,
    add_time: 1543250344,
    up_time: 1543331580,
    __v: 0,
    desc: '',
    markdown: '',
    req_body_type: 'file',
    res_body: '{\n   "x|1-5": [\n      {\n         "email": "@email"\n      }\n   ],\n   "id": "@id",\n   "number": 1\n}',
    req_body_other: 'ffffffffffffffffffffffff',
    tag: [],
    index: 0,
    api_opened: false,
    res_body_is_json_schema: false,
    req_body_form: [
      {
        _id: '5bfd5efca4a6ea4ee3dcb691',
        desc: 'aaa',
        type: 'text',
        name: 'a',
        required: '1',
      },
      {
        _id: '5bfd5efca4a6ea4ee3dcb690',
        desc: 'bbb',
        example: 'B',
        type: 'text',
        name: 'b',
        required: '0',
      },
    ],
    req_body_is_json_schema: true,
    req_params: [],
    req_headers: [],
    req_query: [],
    query_path: {
      path: '/post',
      params: [],
    },
    type: 'static',
    status: 'undone',
    edit_uid: 0,
    username: '方剑成',
  }
  expect(await generateResponsePayloadType(api, 'Test')).toMatchSnapshot()
})

test('json-schema: object', async () => {
  const api: any = {
    _id: 1460,
    method: 'POST',
    catid: 91,
    title: 'post',
    path: '/post',
    project_id: 27,
    res_body_type: 'json',
    uid: 63,
    add_time: 1543250344,
    up_time: 1543331927,
    __v: 0,
    desc: '',
    markdown: '',
    req_body_type: 'file',
    res_body: '{"type":"object","title":"empty object","properties":{"a":{"type":"string","description":"aa"},"b":{"type":"number","description":"bb"},"c":{"type":"object","properties":{"c0":{"type":"boolean","description":"c00"},"c1":{"type":"array","items":{"type":"number","description":"c11111"}}},"required":["c0","c1"]}},"required":["a","b","c"]}',
    req_body_other: 'ffffffffffffffffffffffff',
    tag: [],
    index: 0,
    api_opened: false,
    res_body_is_json_schema: true,
    req_body_form: [
      {
        _id: '5bfd6057a4a6ea4ee3dcb695',
        desc: 'aaa',
        type: 'text',
        name: 'a',
        required: '1',
      },
      {
        _id: '5bfd6057a4a6ea4ee3dcb694',
        desc: 'bbb',
        example: 'B',
        type: 'text',
        name: 'b',
        required: '0',
      },
    ],
    req_body_is_json_schema: true,
    req_params: [],
    req_headers: [],
    req_query: [],
    query_path: {
      path: '/post',
      params: [],
    },
    type: 'static',
    status: 'undone',
    edit_uid: 0,
    username: '方剑成',
  }
  expect(await generateResponsePayloadType(api, 'Test')).toMatchSnapshot()
})

test('json-schema: object, with dataKey', async () => {
  const api: any = {
    _id: 1460,
    method: 'POST',
    catid: 91,
    title: 'post',
    path: '/post',
    project_id: 27,
    res_body_type: 'json',
    uid: 63,
    add_time: 1543250344,
    up_time: 1543331927,
    __v: 0,
    desc: '',
    markdown: '',
    req_body_type: 'file',
    res_body: '{"type":"object","title":"empty object","properties":{"a":{"type":"string","description":"aa"},"b":{"type":"number","description":"bb"},"c":{"type":"object","properties":{"c0":{"type":"boolean","description":"c00"},"c1":{"type":"array","items":{"type":"number","description":"c11111"}}},"required":["c0","c1"]}},"required":["a","b","c"]}',
    req_body_other: 'ffffffffffffffffffffffff',
    tag: [],
    index: 0,
    api_opened: false,
    res_body_is_json_schema: true,
    req_body_form: [
      {
        _id: '5bfd6057a4a6ea4ee3dcb695',
        desc: 'aaa',
        type: 'text',
        name: 'a',
        required: '1',
      },
      {
        _id: '5bfd6057a4a6ea4ee3dcb694',
        desc: 'bbb',
        example: 'B',
        type: 'text',
        name: 'b',
        required: '0',
      },
    ],
    req_body_is_json_schema: true,
    req_params: [],
    req_headers: [],
    req_query: [],
    query_path: {
      path: '/post',
      params: [],
    },
    type: 'static',
    status: 'undone',
    edit_uid: 0,
    username: '方剑成',
  }
  expect(await generateResponsePayloadType(api, 'Test', 'c')).toMatchSnapshot()
})

test('json-schema: array', async () => {
  const api: any = {
    _id: 1460,
    method: 'POST',
    catid: 91,
    title: 'post',
    path: '/post',
    project_id: 27,
    res_body_type: 'json',
    uid: 63,
    add_time: 1543250344,
    up_time: 1543332199,
    __v: 0,
    desc: '',
    markdown: '',
    req_body_type: 'file',
    res_body: '{"type":"array","items":{"type":"object","properties":{"a":{"type":"string","description":"aa"},"b":{"type":"array","items":{"type":"number","description":"bbbb"},"description":"bb"},"c":{"type":"object","properties":{"c0":{"type":"string","description":"c000"}},"required":["c0"]}},"required":["a","b","c"]}}',
    req_body_other: 'ffffffffffffffffffffffff',
    tag: [],
    index: 0,
    api_opened: false,
    res_body_is_json_schema: true,
    req_body_form: [
      {
        _id: '5bfd6167a4a6ea4ee3dcb699',
        desc: 'aaa',
        type: 'text',
        name: 'a',
        required: '1',
      },
      {
        _id: '5bfd6167a4a6ea4ee3dcb698',
        desc: 'bbb',
        example: 'B',
        type: 'text',
        name: 'b',
        required: '0',
      },
    ],
    req_body_is_json_schema: true,
    req_params: [],
    req_headers: [],
    req_query: [],
    query_path: {
      path: '/post',
      params: [],
    },
    type: 'static',
    status: 'undone',
    edit_uid: 0,
    username: '方剑成',
  }
  expect(await generateResponsePayloadType(api, 'Test')).toMatchSnapshot()
})

test('others', async () => {
  const api: any = {
    _id: 1460,
    method: 'POST',
    catid: 91,
    title: 'post',
    path: '/post',
    project_id: 27,
    res_body_type: 'raw',
    uid: 63,
    add_time: 1543250344,
    up_time: 1543332030,
    __v: 0,
    desc: '',
    markdown: '',
    req_body_type: 'file',
    res_body: '22',
    req_body_other: 'ffffffffffffffffffffffff',
    tag: [],
    index: 0,
    api_opened: false,
    res_body_is_json_schema: true,
    req_body_form: [
      {
        _id: '5bfd60bea4a6ea4ee3dcb697',
        desc: 'aaa',
        type: 'text',
        name: 'a',
        required: '1',
      },
      {
        _id: '5bfd60bea4a6ea4ee3dcb696',
        desc: 'bbb',
        example: 'B',
        type: 'text',
        name: 'b',
        required: '0',
      },
    ],
    req_body_is_json_schema: true,
    req_params: [],
    req_headers: [],
    req_query: [],
    query_path: {
      path: '/post',
      params: [],
    },
    type: 'static',
    status: 'undone',
    edit_uid: 0,
    username: '方剑成',
  }
  expect(await generateResponsePayloadType(api, 'Test')).toMatchSnapshot()
})
