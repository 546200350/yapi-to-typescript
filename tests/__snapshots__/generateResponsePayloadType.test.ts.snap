// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`json-schema: array 1`] = `
"export type Test = {
  /**
   * aa
   */
  a: string;
  /**
   * bb
   */
  b: number[];
  c: {
    /**
     * c000
     */
    c0: string;
  };
}[];
"
`;

exports[`json-schema: object 1`] = `
"export interface Test {
  /**
   * aa
   */
  a: string;
  /**
   * bb
   */
  b: number;
  c: {
    /**
     * c00
     */
    c0: boolean;
    c1: number[];
  };
}
"
`;

exports[`json-schema: object, with dataKey 1`] = `
"export interface Test {
  /**
   * c00
   */
  c0: boolean;
  c1: number[];
}
"
`;

exports[`mockjs-json 1`] = `
"export interface Test {
  x?: [
    {
      email?: string;
    }
  ];
  id?: string;
  number?: number;
}
"
`;

exports[`others 1`] = `"export type Test = any"`;
